File: Theory/1. Reversing a String/p1.py
def algo_reverse(my_string):
    new_string = ""
    for char in my_string:
        new_string = char + new_string
    return new_string

my_string = "omae wa mou shindeiru"
print(my_string[::-1])
print(my_string)

print(algo_reverse(my_string))

# No built in function for python.
# This is the same as the algo approach, which is reverse iterate through the string.

File: Theory/1. Reversing a String/p1.java
import java.util.*;

public class p1 {
    public static void main(String[] args) {
        String nani = "omae wa mou shindeiru";
        myReverse(nani);
    }

    public static void myReverse(String theString) {
        // There is no string reverser in Java String primitives.
        // This is because string are immmutable.
        
        StringBuilder newString = new StringBuilder();
        newString = new StringBuilder(theString);

        // This permanently reverses the string.
        newString.reverse();

        System.out.println(newString);

        Scanner input = new Scanner(System.in);
        System.out.println("Enter the string.");
        String inputString = input.nextLine();
        input.close();
        algoReverse(inputString);

    }

    public static void algoReverse(String theString)
    {
        StringBuilder newString = new StringBuilder();
        for (int i = theString.length() - 1; i >= 0; i--)
        {
            newString.append(theString.charAt(i));
        }
        System.out.println(newString);
    }
}